cmd/osbuild-tests/main.go

```diff
@@ -0,0 +1,249 @@
+	}
+	var reply struct {
+		IDs    []deleteUUID  `json:"uuids"`
+		Errors []interface{} `json:"errors"`
+	}
+	rawReply := runCommand(false, "compose", "delete", id.String())
+	err := json.Unmarshal(rawReply, &reply)
+	if err != nil {
+		log.Fatalf("Unexpected reply: " + err.Error())
+	}
+	if len(reply.Errors) != 0 {
+		log.Fatalf("Unexpected errors")
+	}
+	if len(reply.IDs) != 1 {
+		log.Fatalf("Unexpected number of UUIDs returned: %d", len(reply.IDs))
+	}
+	if reply.IDs[0].Status != true {
+		log.Fatalf("Unexpected status %v", reply.IDs[0].Status)
+	}
+}
+
+func waitForCompose(uuid uuid.UUID) string {
+	for {
+		status := getComposeStatus(true, uuid)
+		if status == "FINISHED" || status == "FAILED" {
+			return status
+		}
+		time.Sleep(time.Second)
+	}
+}
+
+func getComposeStatus(quiet bool, uuid uuid.UUID) string {
+	var reply struct {
+		QueueStatus string `json:"queue_status"`
+	}
+	rawReply := runCommand(quiet, "compose", "info", uuid.String())
+	err := json.Unmarshal(rawReply, &reply)
+	if err != nil {
+		log.Fatalf("Unexpected reply: " + err.Error())
+	}
+
+	return reply.QueueStatus
+}
+
+func pushBlueprint(bp *blueprint.Blueprint) {
+	tmpfile, err := ioutil.TempFile("", "osbuild-test-")
+	if err != nil {
+		log.Fatalf("Could not create temporary file: " + err.Error())
+	}
+	defer os.Remove(tmpfile.Name())
+
+	err = toml.NewEncoder(tmpfile).Encode(bp)
+	if err != nil {
+		log.Fatalf("Could not marshapl blueprint TOML: " + err.Error())
+	}
+	if err := tmpfile.Close(); err != nil {
+		log.Fatalf("Could not close toml file: " + err.Error())
+	}
+
+	var reply struct {
+		Status bool `json:"status"`
+	}
+	rawReply := runCommand(false, "blueprints", "push", tmpfile.Name())
+	err = json.Unmarshal(rawReply, &reply)
+	if err != nil {
+		log.Fatalf("Unexpected reply: " + err.Error())
+	}
+	if reply.Status != true {
+		log.Fatalf("Unexpected status %v", reply.Status)
+	}
+}
+
+func deleteBlueprint(bp *blueprint.Blueprint) {
+	var reply struct {
+		Status bool `json:"status"`
+	}
+	rawReply := runCommand(false, "blueprints", "delete", bp.Name)
+	err := json.Unmarshal(rawReply, &reply)
+	if err != nil {
+		log.Fatalf("Unexpected reply: " + err.Error())
+	}
+	if reply.Status != true {
+		log.Fatalf("Unexpected status %v", reply.Status)
+	}
+}
+
+func runCommand(quiet bool, command ...string) json.RawMessage {
+	command = append([]string{"--json"}, command...)
+	cmd := exec.Command("composer-cli", command...)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		log.Fatalf("Colud not create command: " + err.Error())
+	}
+
+	if !quiet {
+		log.Printf("$ composer-cli %s\n", strings.Join(command, " "))
+	}
+
+	cmd.Start()
+
+	var result json.RawMessage
+
+	contents, err := ioutil.ReadAll(stdout)
+	if err != nil {
+		log.Fatalf("Could not read stdout from command: " + err.Error())
+	}
+
+	if len(contents) != 0 {
+		err = json.Unmarshal(contents, &result)
+		if err != nil {
+			// We did not get JSON, try interpreting it as TOML
```

Comment:
There are some commands that output neither `json` nor `toml`, like `status show`. We might need to make the output format an explicit parameter to this function at some point (no need to do it now, though).

Reply:
Yeah, I thought I'd start off trying to make everything JSON and generalize if and when we run into issues. I expect this needs to be refactored eventually, but seems to get us pretty far for now.

Reply:
Yeah. This was (as most of these comments) just me thinking out loud :wink: